docs/en/07-develop/09-udf.md

```diff
@@ -353,15 +390,460 @@ The following table describes the mapping between TDengine SQL data type and Pyt
+Insert OK, 1 row(s) affected (0.003675s)
+
+taos> select ts, v1, v2, v3,  nsum(v1, v2, v3) from t;
+           ts            |     v1      |     v2      |     v3      |     nsum(v1, v2, v3)      |
+================================================================================================
+ 2023-05-01 12:13:14.000 |           1 |           2 |           3 |              14.000000000 |
+ 2023-05-03 08:09:10.000 |           2 |           3 |           4 |              20.000000000 |
+ 2023-05-10 07:06:05.000 |           3 |           4 |           5 |              26.000000000 |
+ 2023-05-25 09:09:15.000 |           6 |        NULL |           8 |                      NULL |
+Query OK, 4 row(s) in set (0.010653s)
+```
+
+#### Sample 4: Utilize 3rd party package
+
+A UDF which accepts a timestamp and output the next closed Sunday. This sample requires to use third party package `moment`, you need to install it firslty.
+
+```shell
+pip3 install moment
+```
+
+Then compose the Python code in /root/udf/nextsunday.py
+
+```python
+import moment
+
+
+def init():
+    pass
+
+
+def destroy():
+    pass
+
+
+def process(block):
+    rows, cols = block.shape()
+    if cols > 1:
+        raise Exception("require only 1 parameter")
+    if not type(block.data(0, 0)) is int:
+        raise Exception("type error")
+    return [moment.unix(block.data(i, 0)).replace(weekday=7).format('YYYY-MM-DD')
+            for i in range(rows)]
+```
+
+UDF framework will map the TDengine timestamp to Python int type, so this function only accepts an integer representing millisecond. process() firstly validates the parameters, then use `moment` to replace the time, format the result and output. 
+
+Create and test the UDF.
+
+```sql
+create function nextsunday as '/root/udf/nextsunday.py' outputtype binary(10) language 'Python';
+```
+
+If your `taosd` is started using `systemd`, you man encounter the error below. Next we will show how to debug.
+
+```sql
+taos> select ts, nextsunday(ts) from t;
+
+DB error: udf function execution failure (1.123615s)
+```
+
+```shell
+ tail -20 taospyudf.log  
+2023-05-25 11:42:34.541 ERROR [1679419] [PyUdf::PyUdf@217] py udf load module failure. error ModuleNotFoundError: No module named 'moment'
+```
+
+This is because `moment` doesn't exist in the default library search path of python UDF, please check the log file `taosdpyudf.log`. 
+
+```shell
+grep 'sys path' taospyudf.log  | tail -1
+```
+
+```text
+2023-05-25 10:58:48.554 INFO  [1679419] [doPyOpen@592] python sys path: ['', '/lib/python38.zip', '/lib/python3.8', '/lib/python3.8/lib-dynload', '/lib/python3/dist-packages', '/var/lib/taos//.udf']
+```
+
+You may find that the default library search path is `/lib/python3/dist-packages` (just for example, it may be different in your system), but `moment` is installed to `/usr/local/lib/python3.8/dist-packages` (for example, it may be different in your system). Then we change the library search path of python UDF.
+
+Check `sys.path`
```

Comment:
Check current `sys.path`, which must include packages you install with pip3 command previously.